File: frontend/lib/models/resource.dart
class Resource {
  final String id;
  final String title;
  final String description;
  final String subject;
  final int semester;
  final String department;
  final String type;
  final String fileUrl;
  final String publicId;
  final Map<String, dynamic> uploadedBy;
  final DateTime createdAt;

  Resource({
    required this.id,
    required this.title,
    required this.description,
    required this.subject,
    required this.semester,
    required this.department,
    required this.type,
    required this.fileUrl,
    required this.publicId,
    required this.uploadedBy,
    required this.createdAt,
  });

  factory Resource.fromJson(Map<String, dynamic> json) {
    return Resource(
      id: json['_id'],
      title: json['title'],
      description: json['description'] ?? '',
      subject: json['subject'],
      semester: json['semester'],
      department: json['department'],
      type: json['type'],
      fileUrl: json['fileUrl'],
      publicId: json['publicId'] ?? '',
      uploadedBy: json['uploadedBy'] is Map ? json['uploadedBy'] : {'name': 'Unknown'},
      createdAt: DateTime.parse(json['createdAt']),
    );
  }
}

File: backend/models/resource.js
import mongoose from "mongoose";

const resourceSchema = new mongoose.Schema({
  title: { type: String, required: true, trim: true },
  description: { type: String, trim: true },
  subject: { type: String, required: true }, // Course Code or Name
  semester: { type: Number, required: true },
  department: { type: String, required: true, enum: ['CSE', 'ECE', 'ME', 'CE', 'EEE', 'MME', 'DSAI', 'BT', 'CHE'] },
  type: { type: String, enum: ['Notes', 'Paper', 'Book', 'Assignment', 'Other'], default: 'Notes' },
  fileUrl: { type: String, required: true },
  publicId: { type: String }, // Cloudinary public_id for deletion
  uploadedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  upvotes: [{ type: mongoose.Schema.Types.ObjectId, ref: 'User' }],
  downloads: { type: Number, default: 0 }
}, { timestamps: true });

const Resource = mongoose.model("Resource", resourceSchema);
export default Resource;
